basic_features/basic_mod_data_checker.py

```diff
@@ -0,0 +1,258 @@
+
+    @staticmethod
+    def match_index_of_pattern(search_str: str, pattern: re.Pattern) -> Optional[int]:
+        """Get the index of the matched group if `search_str` matches `pattern`
+        (from this dict).
+
+        Returns:
+            The 0-based index of the matched group or None for no match.
+        """
+        if not (match := pattern.match(search_str)) or match.lastindex is None:
+            return None
+        return match.lastindex - 1
+
+    @staticmethod
+    def regex_from_glob_list(glob_list: Iterable[str]) -> re.Pattern:
+        """Returns a regex pattern form a list of glob patterns.
+
+        Every pattern has a capturing group, so that `match.lastindex - 1` will
+        give the `glob_list` index.
+        """
+        return re.compile(
+            f'(?:{"|".join(f"({fnmatch.translate(f)})" for f in glob_list)})', re.I
+        )
+
+
+class GlobPatternDict(TypedDict, total=False):
+    """See: `BasicModDataChecker`"""
+
+    unfold: Iterable[str] | None
+    valid: Iterable[str] | None
+    delete: Iterable[str] | None
+    move: MutableMapping[str, str] | None
+
+
+class BasicModDataChecker(mobase.ModDataChecker):
+    """Game feature that is used to check and fix the content of a data tree
+    via simple file definitions.
+
+    The file definitions support glob pattern (without subfolders) and are checked in
+    definition order (passed either as dict or kwargs).
+
+    Args:
+        file_patterns (optional): An dict with the keys below or as kwargs:
+
+        unfold (optional): Folders to unfold (remove and move contents to parent),
+            after being checked and fixed recursively.
+
+        valid (optional): Files and folders in the right path.
+            Check result: `mobase.ModDataChecker.VALID`.
+
+        delete (optional): Files/folders to delete.
+            Check result: `mobase.ModDataChecker.FIXABLE`.
+
+        move (optional): Files/folders to move and their target.
+            If the path ends with `/` or `\\`, the entry will be inserted
+            in the corresponding directory instead of replacing it.
+
+            Check result: `mobase.ModDataChecker.FIXABLE`.
+
+        Example::
+
+            BasicModDataChecker(
+                valid=["valid_folder", "*.ext1"]
+                move={"*.ext2": "path/to/target_folder/"}
+            )
+
+            BasicModDataChecker(
+                {
+                    "valid": ["valid_folder", "*.ext1"],
+                    "move": {"*.ext2": "path/to/target_folder/"}
+                }
+            )
+
+    See Also:
+        `mobase.IFileTree.move`
+    """
+
+    file_patterns = None  # type: GlobPatternDict
+    """Use `update_patterns` for modifications."""
+
+    _regex: RegexPatternDict
+    """The regex patterns derived from the file (glob) patterns."""
+
+    _move_targets: Sequence[str]
+
+    # Overloads as workaround for **kwargs: Unpack[GlobPatternDict]
```

Comment:
That comments do not really make sense... You have fixed `kwargs`, so I don't really see a gain as passing `kwargs` instead of the 4 args, it's more boilerplate code to avoid passing 4 args to a function. There should be a single `__init__` with the second overload signature and simply passing `unfold=unfold`, etc., to `update_patterns`. Same applies to `update_patterns`.

Reply:
Since the order of the `kwargs` is the order in which the checks and fixes are processed (as described in the class docstring), I need to use `kwargs` / `dict`, which preserve the definition order. A `TypedDict` ensures that only valid keys are used, which is why `Unpack` would be preferable, but will only be added in Python 3.11.

We could also add `typing_extensions` as a dev dependency. Then I could remove all the overloads.

Reply:
f3977a9: Removed the kwargs + overloads in favor of a clear distinction between setting (overwriting)  `file_patterns` via `__init__` or `set_patterns` and updating them (changing values while keeping order) via `update_patterns`.